### ISSUE TITLE
Refine Errors

### ISSUE THREAD

**Issue by kazk** [kube-rs/kube-rs#688]
This issue keeps track of tasks related to improving errors in `kube`.

Currently, `kube::Error` is a huge enum containing all the possible errors, and many error variants don't provide any additional context (e.g., `SerdeError(#[from] serde_json::Error)`). Users need to constantly ignore irrelevant error variants, and they don't know which are relevant.

The use of `#[from]` attribute, which derives `impl From<T> for kube::Error`, allows implicit conversion, so the source code doesn't help much, either. The code is slightly more concise with `From`, but the information is lost, allows us to be more sloppy, and not think about error cases as much as we should.

We'll start addressing this problem by removing `From` which will force us to attach context, and gradually breaking down the error into smaller and more specific errors (proposed in #686).

### Tasks

- [x] Remove `impl From<T> for kube::Error` and require explicit error mapping (#686)
- [x] Remove `kube_core::Error` and add more specific errors as needed (#686)
- [ ] Gradually break `kube::Error` down to smaller, and more specific error types.
- [x] **dependency**: Replace `snafu` with `thiserror` in `kube-runtime` and examples (#686)
- [ ] **interoperability**: Change `Display` to follow API Guidelines [`C-GOOD-ERR`](https://rust-lang.github.io/api-guidelines/interoperability.html#c-good-err): "The error message given by the `Display` representation of an error type should be lowercase without trailing punctuation, and typically concise."
- [ ] **interoperability**: Follow the [guidelines for implementing `Display::fmt` and `Error::source`](https://blog.rust-lang.org/inside-rust/2021/07/01/What-the-error-handling-project-group-is-working-towards.html#guidelines-for-implementing-displayfmt-and-errorsource) and remove `source` from `Display` when the ecosystem is ready. Originally discussed in #515.

**Comment by lswith** [comment]
I'd really like for kube::api::Error to also tell me which path is being used in the request. The current printing of 404s is:
```
ApiError: \"404 page not found\\n\": Failed to parse error data (ErrorResponse { status: \"404 Not Found\", message: \"\\\"404 page not found\\\\n\\\"\", reason: \"Failed to parse error data\", code: 404 })"}
```

No idea which resource this is for.

**Comment by clux** [comment]
> I'd really like for kube::api::Error to also tell me which path is being used in the request. The current printing of 404s is:
> 
> ```
> ApiError: \"404 page not found\\n\": Failed to parse error data (ErrorResponse { status: \"404 Not Found\", message: \"\\\"404 page not found\\\\n\\\"\", reason: \"Failed to parse error data\", code: 404 })"}
> ```
> 
> No idea which resource this is for.

I don't believe that is part of the response from the apiserver. We get a [metav1.Stauts](https://github.com/kubernetes/apimachinery/blob/60eaa653342bd0cc48626fa2604bb6d3e7171706/pkg/apis/meta/v1/types.go#L737) which we turn into an [ErrorResponse](https://github.com/kube-rs/kube/blob/c7054b5c86d09947cc4dce210cd24d8487b16f2e/kube-core/src/error.rs#L7) or a [response::Status](https://github.com/kube-rs/kube/blob/c7054b5c86d09947cc4dce210cd24d8487b16f2e/kube-core/src/response.rs#L6C1-L6C1).

The normal way to get this information is to enable tracing (as per examples) and see what is being logged. All request methods log a debug level item about the call. See https://kube.rs/troubleshooting/#request-inspection

**Comment by TilBlechschmidt** [comment]
Just noticed the existing discussion around this topic in https://github.com/kube-rs/kube/pull/515. Given that this was 4 years ago, my question boils down to "is it time to re-evaluate"? While still nightly, `std::error::Report` was merged in January of 2022 and many libraries like `eyre` now rely on `Error::source()` and `Display` which makes using those with `kube` rather unergonomic.

This is more of a personal opinion but it seems that the status quo represents the worst of all variants. Using `Display` produces very repetitive and extremely hard to parse errors (without pasting them into an editor and dissecting them). Using `Report` and similar libraries produces a potentially very long list of error messages which is similarly hard to parse. Producing "clean" error stacks using such tools is far from trivial and as it stands the "source is always a suffix" does not hold true for all variants / errors from `kube` either.

---

~~At the moment, the `Display` implementation of `kube_client::Error` prints the `Error::source` for many variants. This leads to very long, duplicated, and hard to grok outputs when using [`std::error::Report`](https://doc.rust-lang.org/std/error/struct.Report.html), [`eyre`](https://docs.rs/eyre/latest/eyre/struct.Report.html#display-representations), anyhow, or similar.~~

~~Given that other crates like [`hyper`](https://docs.rs/hyper/latest/hyper/struct.Error.html#formatting) follow the pattern of explicitly _not_ including the lower-level source of the error in their formatted output (leaving the choice of how to print the error to the library consumer), would you be open to doing the same for `kube_client::Error` and friends?~~

~~I have created a [very quick draft](https://github.com/kube-rs/kube/commit/b16bfb6a060c1e9974e32b6370bc9e269af9ff54) of how I would imagine it. Definitely needs some polishing and there are other errors types around which follow the same pattern. If you agree with the general approach to printing errors I could create a full PR from it soon™~~

## Example

Here is how a common `Error::source()` trail looks at the moment:

```
1. failed to apply object: Patching TenantControlPlane `kamaji/example/hello-ppi` failed

2. Patching TenantControlPlane `kamaji/example/hello-ppi` failed

3. ApiError: 404 page not found
   : Failed to parse error data (Status { status: Some(Failure), code: 404, message: "404 page not found\n", metadata: None, reason: "Failed to parse error data", details: None })

4. 404 page not found
   : Failed to parse error data
```

- (1) comes from `kube::runtime::finalizer::Error`
- (2) is my custom error type wrapping `kube_client::Error` as its source 
- (3) + (4) is `kube_client::Error` and `Status` respectively

If my error would print its source like `kube` does, the first line would be prohibitively long. The newline originating from the server makes it even harder to interpret than it already is due to the way `Status` prints the response. Might be useful to apply `.trim()` here?
